Respect shell globs when running `docker cp`.

calavera opened the pull request:
Fixes #6013

/cc @huslage, @jlhawn

Signed-off-by: David Calavera david.calavera@gmail.com

duglin commented:
+1 to the idea.
Would it make sense to use go's glob stuff so we support more than just * - like we do in the builder code?

calavera commented:
@duglin can you point me to the builder part you're referring too? I'm slightly lost. I don't know if you saw that the patch uses go's `filepath.Glob` for file expansion, I don't know if that's what you meant.

duglin commented:
@calavera I was just skimming the PR and noticed that https://github.com/docker/docker/pull/12858/files#diff-6ce9e79ddb91a3f06352abe1f2c72ecbR990 only looked for "*" and I think it would need to look for more than that if it was to support the full wildcard set.

duglin commented:
Haven't played with Glob() but I bet you could just use it even in the non-wildcard case and it'll still return a match - then you don't need to check to see if wildcards are used at all.

calavera commented:
> I bet you could just use it even in the non-wildcard case and it'll still return a match

yeah, I thought about that last night, we can probably remove the condition and use glob to match anything. I'll change the code to reflect that.

calavera commented:
@duglin I've unified this logic to always use glob and added test to verify that other matching expressions work as expected. The tests for the old behavior still pass :tada:

duglin commented:
excellent! will review in a little bit....

duglin commented on daemon/container.go:
```diff
@@ -1557,3 +1548,37 @@ func (c *Container) LogDriverType() string {
+	}
+
+	baseDir := filepath.Dir(basePath)
+	for baseDir != "" && !strings.HasPrefix(all[0], baseDir) {
```
Can you explain when you would need to make the baseDir up one level?

calavera replied:
Imagine that the matcher is `/tmp/**/*`. You want to find the base directory common to the expanded paths, for instance `/tmp/foo/test`, and the matcher. Otherwise, you could end up using `/tmp/**` as a base directory, and that doesn't exist.

duglin replied:
hmm if my container has:
```
/a/b/c/d/e/file
```
and I ask for /a/b/c/d/e/file  I get back just 'file'
but if I ask for /a/b/*/*/*/file I get back "c/d/e/file".  
I'm not sure but is this really what should happen?  Should we get different results when using "*" even if it matches just a single path?

Just brainstorming, but I wonder if perhaps instead it should look for the longest common path across all of the files and use that as the root dir.  For example, the result list was:
```
/a/b/c/d/e/file
/a/b/file
```
this would result in a common path of "/a/b" and therefore the paths returns to the CLI would be:
```
c/d/e/file
file
```

calavera replied:
yeah, I think you're right. The problem here is that there are not enough tests to cover other behaviors with more than one file. I'll change the logic to reflect that.

duglin commented on daemon/container.go:
```diff
@@ -1557,3 +1548,37 @@ func (c *Container) LogDriverType() string {
+
+	p := fmt.Sprintf("%s%s", baseDir, string(filepath.Separator))
+	for i, f := range all {
+		r := strings.Replace(f, p, "", -1)
```
I think doing `r := f[len(p):]` would be faster and safer since the `-1` could force the replacement to happen multiple times in for the path, and not just the leading bit.
```
l := len(fmt.Sprintf("%s%s", baseDir,string(filepath.Separator))
for ...
   r := f[l:]
   ...
```

calavera replied:
aggreed

duglin commented on integration-cli/docker_cli_cp_test.go:
```diff
@@ -596,3 +596,66 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
+
+	cleanedContainerID := strings.TrimSpace(out)
+
+	out, _ = dockerCmd(c, "wait", cleanedContainerID)
```
I think it would be easier to just remove the `-d` from the `docker run` and name the container, then use that name instead of the id later on.

calavera replied:
I'm just copying how we initialize other containers for integration tests. I rather not doing anything different.

duglin commented on integration-cli/docker_cli_cp_test.go:
```diff
@@ -596,3 +596,66 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
+		c.Fatal("failed to set up container", out)
+	}
+
+	if err := os.MkdirAll(cpTestPath, os.ModeDir); err != nil {
```
I can't figure out what the "hostFile" is used for - can you explain?
You appear to open it, write to it but don't close it until the defer is run.  But I also can't seem to see you actually using that file later on either.

calavera replied:
It might not be used, I copied the initialization from other cp tests. I'll double check.

duglin commented on integration-cli/docker_cli_cp_test.go:
```diff
@@ -596,3 +596,66 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
+
+		os.RemoveAll(filepath.Join(tmpdir, "*"))
+	}
+}
```
Can we get a test for the case where it glob doesn't match any file?

calavera replied:
totally :+1:

duglin commented on pkg/symlink/fs_test.go:
```diff
@@ -400,3 +400,17 @@ func TestFollowSymlinkNoLexicalCleaning(t *testing.T) {
 	}
 }
+
+func TestFollowSymlinkWildcard(t *testing.T) {
```
Is this new test related to this PR?

calavera replied:
it is. Just to make sure that the symlink library doesn't try to expand glob matches and put them in right context.

calavera commented:
@duglin I've changed the implementation to match the path behavior you described, it makes way more sense. I've also added tests for that behavior and not found paths.

Let me know what you think.

duglin commented on daemon/container.go:
```diff
@@ -1627,3 +1618,53 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
+		if strings.HasPrefix(baseDir, p) {
+			baseDir = p
+		} else if !strings.HasPrefix(p, baseDir) {
+			d1 := filepath.Dir(baseDir)
```
Would you need to keep moving up the tree looking for common paths instead of just one level?  I would think that if baseDir was "/a/b/c" and p was "/d/e/f" then baseDir after this if-statement would need to be just "/".  Right?

duglin commented on daemon/container_unit_test.go:
```diff
@@ -195,3 +196,31 @@ func TestValidContainerNames(t *testing.T) {
+	if b := findSignificantParent(paths); b != "/etc" {
+		t.Fatalf("Expected /etc, was %s\n", b)
+	}
+}
```
Let's add a test for things like:
```
/etc/foo/bar
/tmp/boogie/nights
```
to make sure it comes up with `/`.

And:
```
/tmp/dir1/dir2/dir3/dir4/file
/tmp/dir1/dir3/dir4/dir5/file
```
results in `/tmp/dir1` to make sure we walk up more than one level when needed.

duglin commented on daemon/container.go:
```diff
@@ -1627,3 +1618,53 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
+
+	relative := make([]string, len(paths))
+
+	p := fmt.Sprintf("%s%s", baseDir, string(filepath.Separator))
```
I think you need an if-statement around the adding of "/" so that if baseDir is just "/" you don't add another one.

duglin commented:
I haven't been able to figure out what's going on but I do see something kind of neat/funky  :-)
```
$curl -s -H "Content-Type:application/json" -X POST -d '{"Resource":"/*"}' http://127.0.0.1:2376/v1.19/containers/bee37b85c512/copy  | tar -tf- | head -10
.dockerenv
.dockerinit
a/
a/b/
a/b/c/
a/b/c/d/
a/b/c/d/e/
a/b/c/d/e/file
...

But then look at this:
$curl -s -H "Content-Type:application/json" -X POST -d '{"Resource":"/"}' http://127.0.0.1:2376/v1.19/containers/bee37b85c512/copy  | tar -tf- | head -10
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/.dockerenv
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/.dockerinit
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/a/
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/a/b/
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/a/b/c/
bee37b85c512f864f46b01c12bf33e3e80cbb286a3b91ef69051c03b788d27a8/a/b/c/d/
...
```
I don't know for sure but I'm guessing this might have something to do with this line:
https://github.com/docker/docker/pull/12858/files?diff=unified#diff-6ce9e79ddb91a3f06352abe1f2c72ecbR1645
and then passing in "" as an IncludeFile into the archiver.  But didn't dig any deeper yet....

duglin commented on daemon/container.go:
```diff
@@ -1627,3 +1618,53 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
+	for i, f := range paths {
+		r := f[len(p):]
+
+		if r != "" {
```
When would we expect r to be "" ?

calavera commented:
@duglin I decided to go with another implementation(a pseudo-tree) to find the right path and it looks like it works properly. I've added more test cases that should cover the scenarios you describe.

duglin commented on daemon/container.go:
```diff
@@ -1627,3 +1618,93 @@ func (container *Container) monitorExec(execConfig *execConfig, callback execdri
+	return &archiveFilter{baseDir, relative}, nil
+}
+
+func findSignificantParent(paths []string) string {
```
This seems like a lot of work.  Would something a little smaller like this work?  https://gist.github.com/duglin/c88c3d5fde9c9ac6fd22

calavera committed:
Respect shell globs when running `docker cp`.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
@duglin it looks like everything works as expected with your patch :tada:

duglin commented on integration-cli/docker_cli_cp_test.go:
```diff
@@ -596,3 +596,82 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
+		c.Fatal("failed to set up container", out)
+	}
+
+	if err := os.MkdirAll(cpTestPath, os.ModeDir); err != nil {
```
Is the stuff on lines 615 thru 619 needed?  I'm not sure why we would need to create these dirs on the local machine if we cp into "tmpdir" ?

duglin commented on integration-cli/docker_cli_cp_test.go:
```diff
@@ -596,3 +596,82 @@ func (s *DockerSuite) TestCpNameHasColon(c *check.C) {
+				c.Errorf("output doesn't match the input -- %s - %s\n", cs.pathMatch, cs.tmpName)
+			}
+		}
+		os.RemoveAll(filepath.Join(tmpdir, "*"))
```
Is this needed if we have the `defer os.RemoveAll(tmpdir)` above?

calavera replied:
yes, it cleans the directory between test cases.

duglin replied:
ok

duglin replied:
Are you sure "*" is ok in the path? I didn't notice the docs saying it deal with globs.  You might be better off just erasing all of tmpdir and then recreating it before each iteration.

duglin replied:
I meant "*" in the RemoveAll arg.

calavera commented:
I've cleaned up that dead code in the test.

calavera commented:
@duglin it looks like your algorithm didn't handle copying `/` correctly either, but it was caused for a previous cleanup that we do in the input. I'm generating temp directories for every test case now and it looks like it works as expected.

duglin commented on daemon/container.go:
```diff
@@ -1034,28 +1034,23 @@ func (container *Container) Copy(resource string) (io.ReadCloser, error) {
 		}
 	}()
 
+	if len(resource) == 0 { // Copying `/` we expand to copy its content
```
Not sure, but are you sure we wouldn't be better off generating an error? Its like a "cp" w/o a src path - kind of odd. And some people may think that "" should mean "current dir" and not "/".

duglin replied:
If we do end up keeping this then let's add a test for it in TestCpGlob()

calavera replied:
this actually is the case when you do `docker cp ID:/ /tmp`, copying the whole file system. We strip the first `/` at the api level and arrive to this point with an empty string. Docker already fails if you try to copy an empty string, but I've added a test case to be explicit about it.

duglin replied:
hmm didn't realize the api layer did that - I kind of wonder why.  Oh well.

duglin commented:
just one comment - otherwise its looking good

duglin commented:
LGTM

jlhawn commented:
I haven't reviewed this yet, but I expect a major conflict will occur with https://github.com/docker/docker/pull/13171 :sob:

calavera commented:
@jlhawn conflicts are always solvable. This only cares about the source, regardless you're copying to the host or other container.

calavera commented:
@jlhawn does your PR implements this feature then? Let me know to close this, I don't want to have it around if it's duplicated.

duglin commented:
Right now it doesn't but that would be really cool if it did - hint hint :-)

LK4D4 commented:
@calavera would you mind to rebase?

calavera commented:
@LK4D4 rebased

icecrime commented:
LGTM

thaJeztah commented on docs/reference/commandline/cli.md:
```diff
@@ -976,6 +976,9 @@ relative to the root of the container's filesystem.
 
+This command respects shell globs inside the container's filesystem. For instance:
+
+    docker cp CONTAINER:/tmp/* /tmp
```
I think wildcards may be a more widely used term than "shell globs". Perhaps;

```Markdown
The `docker cp` command supports the use of wildcards to specify which
files and folders to copy from the container's filesystem.

For example, run this command to copy all files and folders in the
`/tmp/` directory inside the container to the `/home/foo/` directory
on the host:

    $ docker cp CONTAINER:/tmp/* /home/foo/
```

thaJeztah commented:
Thanks @calavera! I added a comment inline. Can you also update the [man/docker-cp.1.md](https://github.com/docker/docker/blob/master/man/docker-cp.1.md) file to include this information?

ping @fredlf @jamtur01 @moxiegirl @SvenDowideit

duglin commented:
I'm a bit confused. If we merged this, and then merged #13171 won't we regress in function?  @jlhawn are you planning on adding glob support soon?

jlhawn commented:
I would prefer to add it separately, as it would make #13171 even larger than it already is.

Also, I'm not 100% on board with the current design of this feature. I'm not a fan of overloading the `copy`/`archive-path` endpoints with the ability to handle globs, perform a search for the most significant parent, and return an archive of *that* common parent directory with all of the relative paths included. This means that the resulting archive will have entries with arbitrary paths depending on the common parent and it doesn't work with what I have currently going on in #13171.

I have suggested another approach where we add a separate endpoint that returns paths in the container which matches a shell glob. The `cp` command in the client can this to expand the source argument. Then all we need to do is implement `docker cp` with multiple sources.

jlhawn commented:
@duglin @icecrime @calavera Should I go ahead and add what I described to #13171?

calavera commented:
I'm closing this PR. I don't want to be biased on my review of #13171.

duglin commented:
I'm ok with not doing the "most significant parent" thing (and basing the results on / instead), but I'm not sure I understand the value of having a separate endpoint just to expand wildcards and then forcing the client to have to issue a second command. 

By doing that you're asking the client to inspect the input, look for all possible wildcards (and deal with escaping), and then do extra processing based on the outcome of that analysis.  What makes it more tricky is that if we decide to add support for a different set of wildcards then each and every client would need to change to match it.  

Unless, you're expecting all callers to first call the new "expand globs API" and then use that result in the "copy files" API? In which case, if calling "expand" is (pretty much) required, why not just always do it under "copy files" for people?  If people really have usecases for not expanding wildcards then we can add a flag to disable it.  Just doesn't seem a user friendly to split it to me.

jlhawn commented:
> Unless, you're expecting all callers to first call the new "expand globs API" and then use that result in the "copy files" API? In which case, if calling "expand" is (pretty much) required, why not just always do it under "copy files" for people?

Yes, I was thinking of *always* having the server expand wildcards. I don't think the API should be a 'copy' API, but it should provide the needed combination of endpoints to make it possible. The added round trips for the multiple API calls is negligible, in my opinion.

> If people really have usecases for not expanding wildcards then we can add a flag to disable it.

Or have users of the `docker cp` command properly escape any glob characters that they don't want the server to expand. This is also what `scp` requires.

> Just doesn't seem a user friendly to split it to me.

User friendly for API users or for CLI users? I think that API users would appreciate the simplicity and understand the separation of functionality. And CLI users would appreciate the familiarity of `docker cp` to `scp`.

jlhawn commented:
I also want to point out that my original PR for `docker cp` (back in January) worked the other way: it actually **was** a copy API, but several of the design reviewers didn't like that approach because it required the server to have too much extra context about the source (when copying to a container) or destination (when copying from a container). I agreed with the arguments that were made and those discussions resulted in the current design with a much simpler archive/extract API where the client would be required to prepare the archives for the desired copy behavior.

KES777 commented:
can not copy from container:

```
$ sudo docker cp maitre_d_postgres:'/opt/monkeyman/bin/ur_2/*5*' ./
Error response from daemon: lstat /var/lib/docker/overlay2/668f3659d6e5212c815170067d8b66ab553355b26ee551273139ecdf387f10f3/merged/opt/monkeyman/bin/ur_2/*5*: no such file or directory
```
